Cannot cast TypeBinding$LocalTypeBinding to IVariableBinding exception on `completionItem/resolve`

Version: latest master (05b938fc61c461e82dc749e705a365ec787eb9fd)
Java 16.0.1


Got the following error when invoking `completionItem/resolve` with the following item:

```
{
  data = {
    decl_signature = "Lio.crate.expression.operator.LikeOperators$CaseSensitivity;",
    name = "SENSITIVE",
    pid = "0",
    rid = "4",
    uri = "file:///path/to/dev/crate/crate/server/src/main/java/io/crate/lucene/LuceneQueryBuilder.java"
  },
  detail = "LikeOperators.CaseSensitivity.SENSITIVE : LikeOperators.CaseSensitivity",
  insertText = "SENSITIVE",
  insertTextFormat = 2,
  kind = 20,
  label = "SENSITIVE : LikeOperators.CaseSensitivity",
  sortText = "999998698",
  textEdit = {
    newText = "SENSITIVE",
    range = {
      end = {
        character = 73,
        line = 294
      },
      start = {
        character = 71,
        line = 294
      }
    }
  }
}
```

The `LuceneQueryBuilder` is part of [crate/crate](https://github.com/crate/crate) on [this commit](https://github.com/crate/crate/commit/2f0724f88c001f402c950253d0b59e534f8e86fa)



```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:120)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:261)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:190)
	at org.eclipse.jdt.ls.core.internal.ParentProcessWatcher.lambda$0(ParentProcessWatcher.java:123)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:831)
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:78)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.util.concurrent.CompletionException: java.lang.ClassCastException: class org.eclipse.jdt.core.dom.TypeBinding$LocalTypeBinding cannot be cast to class org.eclipse.jdt.core.dom.IVariableBinding (org.eclipse.jdt.core.dom.TypeBinding$LocalTypeBinding and org.eclipse.jdt.core.dom.IVariableBinding are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @13acc999)
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:295)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassCastException: class org.eclipse.jdt.core.dom.TypeBinding$LocalTypeBinding cannot be cast to class org.eclipse.jdt.core.dom.IVariableBinding (org.eclipse.jdt.core.dom.TypeBinding$LocalTypeBinding and org.eclipse.jdt.core.dom.IVariableBinding are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @13acc999)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.computeFieldConstantFromTypeAST(JDTUtils.java:1330)
	at org.eclipse.jdt.ls.core.internal.JDTUtils.getConstantValue(JDTUtils.java:1221)
	at org.eclipse.jdt.ls.core.internal.handlers.CompletionResolveHandler.resolve(CompletionResolveHandler.java:198)
	at org.eclipse.jdt.ls.core.internal.handlers.JDTLanguageServer.lambda$6(JDTLanguageServer.java:544)
	at org.eclipse.jdt.ls.core.internal.BaseJDTLanguageServer.lambda$0(BaseJDTLanguageServer.java:75)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)
	... 6 more
```


https://user-images.githubusercontent.com/38700/130791775-bcbc72c8-03e0-4e99-8318-ec53b98b8291.mp4

--- a/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java
+++ b/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JDTUtils.java
@@ -1326,10 +1326,8 @@ public final class JDTUtils {
 		} catch (OperationCanceledException e) {
 			return null;
 		}
-
-		IVariableBinding variableBinding = (IVariableBinding) createBindings[0];
-		if (variableBinding != null) {
-			return variableBinding.getConstantValue();
+		if (createBindings[0] instanceof IVariableBinding) {
+			return ((IVariableBinding) createBindings[0]).getConstantValue();
 		}
 
 		return null;

--- a/org.eclipse.jdt.ls.tests/projects/eclipse/hello/src/org/sample/Enums.java
+++ b/org.eclipse.jdt.ls.tests/projects/eclipse/hello/src/org/sample/Enums.java
@@ -1,0 +1,25 @@
+package org.sample;
+
+public class Enums {
+
+    public enum MyEnum {
+        ENUM1 {
+
+            @Override
+            public String get() {
+                return "ENUM1";
+            }
+        },
+        ENUM2 {
+
+            @Override
+            public String get() {
+                return "ENUM2";
+            }
+        };
+
+        public abstract String get();
+
+    }
+    
+}

--- a/org.eclipse.jdt.ls.tests/projects/eclipse/hello/src/org/sample/TestEnum.java
+++ b/org.eclipse.jdt.ls.tests/projects/eclipse/hello/src/org/sample/TestEnum.java
@@ -1,0 +1,7 @@
+package org.sample;
+
+import org.sample.Enums.MyEnum;
+
+public class TestEnum {
+	private MyEnum myEnum = MyEnum.ENUM1;
+}

--- a/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/HoverHandlerTest.java
+++ b/org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/HoverHandlerTest.java
@@ -685,6 +685,19 @@ public class HoverHandlerTest extends AbstractProjectsManagerBasedTest {
 		assertMatches(expectedJavadoc, javadoc);
 	}
 
+	// https://github.com/eclipse/eclipse.jdt.ls/issues/1856
+	@Test
+	public void testEnum() throws Exception {
+		String payload = createHoverRequest("src/org/sample/TestEnum.java", 5, 33);
+		TextDocumentPositionParams position = getParams(payload);
+		Hover hover = handler.hover(position, monitor);
+		assertNotNull(hover);
+		assertNotNull(hover.getContents());
+		MarkedString signature = hover.getContents().getLeft().get(0).getRight();
+		assertEquals("Unexpected hover " + signature, "java", signature.getLanguage());
+		assertEquals("Unexpected hover " + signature, "ENUM1", signature.getValue());
+	}
+
 	private String getTitleHover(ICompilationUnit cu, int line, int character) {
 		// when
 		Hover hover = getHover(cu, line, character);

